Biopete/gcs cloudsql direct [risk=low]

biopete opened the pull request:
1. Adds generate-cloudsql-db goal to directly make a cloudsql database from a bucket of csvs. See Readme.
2. Moves the make-bq-denormalized-tables into its own goal out of the generate-cdr-counts goal
3. Optimizations, cleanup, and tweaks so old sql dump system still works for test data

freemabd1 commented on README.md:
```diff
@@ -285,10 +285,22 @@ Description of arguments for these scripts are as follows. See examples below.
+#### Generate local mysql databases -- cdr and public for data generated above if you need to develop with a full test database
 `./project.rb generate-local-count-dbs --cdr-version 20180206 --bucket all-of-us-workbench-private-cloudsql`
 ##### Result is
 1. Local mysql database cdr20180206 fully populated with count data from cdr version 20180206
```
Can we add some documentation in the README about how to run db-make-bq-denormalized-tables?

biopete replied:
Yes.

calbach commented on api/docker-compose.yaml:
```diff
@@ -133,6 +151,24 @@ services:
       - GOOGLE_APPLICATION_CREDENTIALS=/w/sa-key.json
     env_file:
       - db-cdr/vars.env
+  db-generate-cloudsql-db:
```
Note for follow-up - I don't think you need 1 entry here per shell script , the configuration across all of these CDR commands are identical save for the script. I think you could probably get away with just one service here like "cdr-scripts", then invoke it like:

```
docker-compse run cdr-script my-script.sh
```

at the callsites. Possibly one additional target needed for those that access a local DB, e.g. cdr-local-scripts. I'd keep this as-is for this PR for now though.

calbach commented on api/libproject/devstart.rb:
```diff
@@ -641,6 +641,17 @@ def run_local_public_data_migrations()
+Common.register_command({
+  :invocation => "make-bq-denormalized-tables",
+  :description => "make-bq-denormalized-tables --bq-project <PROJECT> --bq-dataset <DATASET>
+Generates big query denormalized tables. Used by cohort reviewer. Must be run once when a new cdr is released",
```
reviewer --> builder?

calbach commented on api/db-cdr/generate-cdr/generate-cloudsql-db.sh:
```diff
@@ -0,0 +1,83 @@
+--bucket <BUCKET>"
+
+while [ $# -gt 0 ]; do
+  echo "1 is $1"
```
rm?

calbach commented on api/db-cdr/generate-cdr/generate-cloudsql-db.sh:
```diff
@@ -0,0 +1,83 @@
+
+if [ -z "${BUCKET}" ]
+then
+  #echo "Usage: $USAGE"
```
Uncomment or rm

calbach commented on api/db-cdr/generate-cdr/generate-cloudsql-db.sh:
```diff
@@ -0,0 +1,83 @@
+echo "Start $startDate Stop: $stopDate"
+echo $(date) " Finished generate-clousdsql-db "
+
+exit 0
```
nit:I'm not sure this is necessary, isn't this implicit at the end of a bash script?

calbach commented on api/db-cdr/generate-cdr/cloudsql-import-bucket.sh:
```diff
@@ -0,0 +1,143 @@
+USAGE="./generate-clousql-cdr/cloudsql-import-bucket.sh --project <PROJECT> --instance <INSTANCE> --bucket <BUCKET>"
+# example account for test : all-of-us-workbench-test@appspot.gserviceaccount.com
+while [ $# -gt 0 ]; do
+  echo "1 is $1"
```
rm?

calbach commented on api/db-cdr/generate-cdr/cloudsql-import-bucket.sh:
```diff
@@ -0,0 +1,143 @@
+# Trim leading whitespace from sql service account
+SQL_SERVICE_ACCOUNT=${SQL_SERVICE_ACCOUNT//[[:blank:]]/}
+
+echo "Granting GCS access to ${SQL_SERVICE_ACCOUNT} to bucket $BUCKET/*"
```
Interesting, I thought the import process did this for you. Or is that just the Cloud Console UI that does it?

calbach commented on api/db-cdr/generate-cdr/cloudsql-import-bucket.sh:
```diff
@@ -0,0 +1,143 @@
+CREATE_DUMP_FILE=
+
+# get options
+USAGE="./generate-clousql-cdr/cloudsql-import-bucket.sh --project <PROJECT> --instance <INSTANCE> --bucket <BUCKET>"
```
Could you include usage for --create-dump-file? Looks like you're expecting that to be a string - consider also renaming it since it sound s like a boolean flag currently.

calbach replied:
Ah, I was very mislead by the name I guess. I thought this was actually going to create a dump file, now the functionality makes sense.

I'd say something like `--create-db-from-sql-dump` or `--init-db-from-sql-dump` to make the workflow clearer.

biopete commented:
Does replying to a comment in email work good on GitHub ? That would be awesome.  @freemabd1   I would say not deprecating.  Just we would not use them for a typical use case of making a new cloud sql cdr.  

Generate local count dbs would solve objective 3 too — I want a local full dB for development.  

Basically if you want the dB in the cloud use generate cloud dB.  Otherwise do some other method.

MySQL local dump dB is good if you need to share a dump with someone to try out your code too.  

Peter

> On Sep 17, 2018, at 12:15 PM, Brian Freeman <notifications@github.com> wrote:
> 
> @freemabd1 commented on this pull request.
> 
> So Peter will we be deprecating the following scripts?
> 
> generate-local-count-dbs
> mysqldump-local-db
> In README.md:
> 
> >  3. Browse csvs in browser like here :https://console.cloud.google.com/storage/browser?project=all-of-us-workbench-test&organizationId=394551486437
>  3. Note cdr-version can be ''  to make datasets named cdr and public
> -#### Generate local mysql databases -- cdr and public for data generated above
> +#### Generate cloudsql databases from a bucket without downloading the data
> +`# Once for private cdr`
> +
> +`./project.rb generate-cloudsql-db --project all-of-us-workbench-test --instance workbenchmaindb --database cdr20180913 --bucket all-of-us-workbench-private-cloudsql/cdr20180913`
> +
> +`# Once for public cdr`
> +
> +`./project.rb generate-cloudsql-db --project all-of-us-workbench-test --instance workbenchmaindb --database public20180913 --bucket all-of-us-workbench-private-cloudsql/public20180913`
> +##### Result is
> +1. Databases are live on cloudsql.
> +2. TODO - Run .project.rb update-cdr-versions to use them in your environment ??
> +
> +#### Generate local mysql databases -- cdr and public for data generated above if you need to develop with a full test database
>  `./project.rb generate-local-count-dbs --cdr-version 20180206 --bucket all-of-us-workbench-private-cloudsql`
>  ##### Result is
>  1. Local mysql database cdr20180206 fully populated with count data from cdr version 20180206
> Can we add some documentation in the README about how to run db-make-bq-denormalized-tables?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

calbach commented on api/db-cdr/generate-cdr/cloudsql-import-bucket.sh:
```diff
@@ -0,0 +1,143 @@
+#!/bin/bash
+
+# Imports all sql and csv files in the bucket to the database.
+# IF --create-dump-file is specified, it is ran first
```
I'm not following - what is the relationship between --create-dump-file and the rest of this script? It seems almost like an unrelated action

danqrodney commented on api/db-cdr/generate-cdr/cloudsql-import-bucket.sh:
```diff
@@ -0,0 +1,143 @@
+    gsutil mv $create_gs_file gs://$BUCKET/imported_to_cloudsql/
+fi
+
+#Import files, do sql first as they are intended for schema changes and such
```
So to clarify: we are still using SQL dumps for the schema portion, but are populating the tables via CSV files?

biopete replied:
Yes. Liquibase still making the db.

biopete commented on api/src/dev/server/Dockerfile:
```diff
@@ -10,7 +10,7 @@ RUN apk --no-cache add openjdk8
 #
 
-ENV CLOUD_SDK_VERSION 165.0.0
+ENV CLOUD_SDK_VERSION 216.0.0
```
Note this sdk update ... in case you see something that is deprecated and breaks. I tried to find what version gcloud sql import csv was released out of beta into the official sdk but i could not find it.

danqrodney replied:
OK, please notify #workbench-eng.

biopete commented:
I think this is ready for review again. I rebased and Srushti is generating databases in test which I think then we will add as as another cdr release to test the whole system of releasing a new cdr to the environments.  Thanks.

SrushtiGangireddy commented on README.md:
```diff
@@ -281,14 +281,30 @@ Description of arguments for these scripts are as follows. See examples below.
+1. Databases are live on cloudsql.
+2. TODO - Run .project.rb update-cdr-versions to use them in your environment ??
+
+#### Generate local mysql databases -- cdr and public for data generated above if you need to develop with a full test database
```
Data generated above seems to be lil confusing. May be we can change it to "from data in csvs in buckets" or whatever.

danqrodney commented on README.md:
```diff
@@ -281,14 +281,30 @@ Description of arguments for these scripts are as follows. See examples below.
+`./project.rb generate-cloudsql-db --project all-of-us-workbench-test --instance workbenchmaindb --database public20180913 --bucket all-of-us-workbench-private-cloudsql/public20180913`
+##### Result is
+1. Databases are live on cloudsql.
+2. TODO - Run .project.rb update-cdr-versions to use them in your environment ??
```
Please update this at some point

danqrodney commented on api/db-cdr/generate-cdr/cloudsql-import.sh:
```diff
@@ -34,50 +44,133 @@ then
+        echo "Importing file into $table. If table does not exist in the database, this file is moved to imported."
+        gcloud sql import csv $INSTANCE $gs_file --project $PROJECT --quiet --account $SERVICE_ACCOUNT \
+        --database=$DATABASE --table=$table --async
+        import_wait $gs_file 15
```
I would have thought large CSV files would take much longer than 15 seconds to import?

danqrodney replied:
Right, and I was saying: once you poll every second (which I thought you were doing), it's fine to set a bigger timeout... under normal circumstances you won't hit it.

Anyway, this is fine for now, but if we start noticing occasional failures due to slow imports, we could consider bumping these up.

danqrodney commented on api/db-cdr/generate-cdr/cloudsql-import.sh:
```diff
@@ -34,50 +44,133 @@ then
+        echo "Skipping create db sql file ";
+    else
+        gcloud sql import sql $INSTANCE $gs_file --project $PROJECT --account $SERVICE_ACCOUNT --quiet --async
+        import_wait $gs_file 15
```
15 seconds seems a little skimpy. Does it hurt to make this say 60?

danqrodney commented on api/docker-compose.yaml:
```diff
@@ -115,6 +115,24 @@ services:
       command: wait-for db:3306 -- ./generate-cdr/init-new-cdr-db.sh --cdr-db-name public --run-list data --context local
       env_file:
         - db-cdr/vars.env
+  db-make-bq-denormalized-tables:
```
Does this have to be a separate step? Can't we combine this with the other generate CDR BQ logic?

biopete commented:
Ok. I updated the readme. Handled all the concerns. Made cloudsql-import more robust by erroring if the database doesn't exist after you try to create it or try to import csvs to it. gcloud sql import csv does not fail if you try to import it to a database that does not exist . lol. I'm not real sure what it is doing when that happens lol. 

I'm rebasing and testing out some other processes in the readme. The cdr20180920 and public20180920 are built in cloudsql . I am exporting a dump of them to mysql_dumps/cdr_public20180920 to import locally. 

Also, i updated the cdr_versions_test.json with a new cdr ... I left the cdrDbName the current one so i wouldn't break anything. 

Any further comments or concerns ?
